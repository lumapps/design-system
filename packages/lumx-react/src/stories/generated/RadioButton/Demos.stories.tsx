/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/radio-button/Demos' };

export { App as Default } from './default';
