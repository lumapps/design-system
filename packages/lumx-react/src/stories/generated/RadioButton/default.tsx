../../../../../site-demo/content/product/components/radio-button/react/default.tsx