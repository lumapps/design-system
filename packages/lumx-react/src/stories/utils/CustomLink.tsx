/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
