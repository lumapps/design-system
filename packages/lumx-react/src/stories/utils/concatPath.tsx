export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
