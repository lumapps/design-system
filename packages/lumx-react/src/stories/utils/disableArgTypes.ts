export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
