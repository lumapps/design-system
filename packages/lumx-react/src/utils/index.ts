/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
