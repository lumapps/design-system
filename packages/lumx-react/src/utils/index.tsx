export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
