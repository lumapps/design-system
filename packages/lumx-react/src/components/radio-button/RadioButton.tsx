import React, { useMemo, forwardRef, ReactNode, SyntheticEvent } from 'react';

import classNames from 'classnames';
import { uid } from 'uid';

import { InputHelper, InputLabel, Theme } from '@lumx/react';

import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

/**
 * Defines the props of the component.
 */
export interface RadioButtonProps extends GenericProps {
    /** Helper text. */
    helper?: string;
    /** Native input id property. */
    id?: string;
    /** Whether it is checked or not. */
    isChecked?: boolean;
    /** Whether the component is disabled or not. */
    isDisabled?: boolean;
    /** Label content. */
    label?: ReactNode;
    /** Native input name property. */
    name?: string;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** Native input value property. */
    value?: string;
    /** On change callback. */
    onChange?(value?: string, name?: string, event?: SyntheticEvent): void;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'RadioButton';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<RadioButtonProps> = {
    theme: Theme.light,
};

/**
 * RadioButton component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const RadioButton: Comp<RadioButtonProps, HTMLDivElement> = forwardRef((props, ref) => {
    const {
        checked,
        className,
        disabled,
        helper,
        id,
        isChecked = checked,
        isDisabled = disabled,
        label,
        name,
        onChange,
        theme,
        value,
        ...forwardedProps
    } = props;
    const radioButtonId = useMemo(() => id || `${CLASSNAME.toLowerCase()}-${uid()}`, [id]);

    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (onChange) {
            onChange(value, name, event);
        }
    };

    return (
        <div
            ref={ref}
            {...forwardedProps}
            className={classNames(
                className,
                handleBasicClasses({
                    isChecked,
                    isDisabled,
                    isUnchecked: !isChecked,
                    prefix: CLASSNAME,
                    theme,
                }),
            )}
        >
            <div className={`${CLASSNAME}__input-wrapper`}>
                <input
                    className={`${CLASSNAME}__input-native`}
                    disabled={isDisabled}
                    id={radioButtonId}
                    tabIndex={isDisabled ? -1 : 0}
                    type="radio"
                    name={name}
                    value={value}
                    checked={isChecked}
                    onChange={handleChange}
                />

                <div className={`${CLASSNAME}__input-placeholder`}>
                    <div className={`${CLASSNAME}__input-background`} />
                    <div className={`${CLASSNAME}__input-indicator`} />
                </div>
            </div>

            <div className={`${CLASSNAME}__content`}>
                {label && (
                    <InputLabel htmlFor={radioButtonId} theme={theme} className={`${CLASSNAME}__label`}>
                        {label}
                    </InputLabel>
                )}
                {helper && (
                    <InputHelper theme={theme} className={`${CLASSNAME}__helper`}>
                        {helper}
                    </InputHelper>
                )}
            </div>
        </div>
    );
});
RadioButton.displayName = COMPONENT_NAME;
RadioButton.className = CLASSNAME;
RadioButton.defaultProps = DEFAULT_PROPS;
