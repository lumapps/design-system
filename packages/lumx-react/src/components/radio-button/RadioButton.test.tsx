import React, { InputHTMLAttributes, LabelHTMLAttributes, ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';

import { Theme } from '@lumx/react';
import { RadioButton, RadioButtonProps } from './RadioButton';

const CLASSNAME = RadioButton.className as string;

type SetupProps = Partial<RadioButtonProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propsOverride: SetupProps = {}, shallowRendering = true) => {
    const props: any = { id: 'fixedId', ...propsOverride };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: Wrapper = renderer(<RadioButton {...props} />);

    return {
        helper: wrapper.find(`.${CLASSNAME}__helper`),
        input: wrapper.find(`input`),
        label: wrapper.find(`.${CLASSNAME}__label`),
        props,
        wrapper,
    };
};

describe(`<${RadioButton.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render defaults', () => {
            const { wrapper, input, helper } = setup({});
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();

            expect(input).toExist();
            const inputProps: InputHTMLAttributes<HTMLInputElement> = input.props();
            expect(inputProps.checked).toBeFalsy();
            expect(inputProps.disabled).toBeFalsy();

            expect(helper).not.toExist();
        });

        it('should render checked, disabled & id', () => {
            const props = {
                isChecked: true,
                isDisabled: true,
                id: 'inputID',
            };
            const { wrapper, input } = setup(props);
            expect(wrapper).toMatchSnapshot();

            const inputProps: InputHTMLAttributes<HTMLInputElement> = input.props();
            expect(inputProps.checked).toBe(true);
            expect(inputProps.disabled).toBe(true);
            expect(inputProps.id).toEqual(props.id);
        });

        it('should render label & helper', () => {
            const props = {
                helper: 'Helper',
                label: 'Label',
            };
            const { wrapper, label, helper, input } = setup(props as Partial<RadioButtonProps>);
            expect(wrapper).toMatchSnapshot();

            expect(input).toExist();
            const inputProps: InputHTMLAttributes<HTMLInputElement> = input.props();
            expect(inputProps.checked).toBeFalsy();
            expect(inputProps.disabled).toBeFalsy();

            expect(label).toExist();
            expect(label.contains(props.label)).toBe(true);
            const labelProps: LabelHTMLAttributes<HTMLLabelElement> = label.props();
            expect(labelProps.htmlFor).toEqual(inputProps.id);

            expect(helper).toExist();
            expect(helper.contains(props.helper)).toBe(true);
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use props for classes', () => {
            const props = {
                isChecked: true,
                isDisabled: true,
                theme: Theme.dark,
            };
            const { wrapper } = setup(props as RadioButtonProps);

            expect(wrapper).toHaveClassName(CLASSNAME);
            expect(wrapper).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'isDisabled', value: true }));
            expect(wrapper).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'isChecked', value: true }));
            expect(wrapper).not.toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'isUnchecked', value: true }));
            expect(wrapper).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'theme', value: props.theme }));
        });
    });

    // 3. Test events.
    describe('Events', () => {
        // Nothing to do here.
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        // Nothing to do here.
    });

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    // Common tests suite.
    commonTestsSuite(setup, { prop: 'wrapper', className: 'wrapper' }, { className: CLASSNAME });
});
