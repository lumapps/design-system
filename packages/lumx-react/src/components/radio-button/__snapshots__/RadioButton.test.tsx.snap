// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButton> Snapshots and structure should render checked, disabled & id 1`] = `
<div
  className="lumx-radio-button lumx-radio-button--is-checked lumx-radio-button--is-disabled lumx-radio-button--theme-light"
>
  <div
    className="lumx-radio-button__input-wrapper"
  >
    <input
      checked={true}
      className="lumx-radio-button__input-native"
      disabled={true}
      id="inputID"
      onChange={[Function]}
      tabIndex={-1}
      type="radio"
    />
    <div
      className="lumx-radio-button__input-placeholder"
    >
      <div
        className="lumx-radio-button__input-background"
      />
      <div
        className="lumx-radio-button__input-indicator"
      />
    </div>
  </div>
  <div
    className="lumx-radio-button__content"
  />
</div>
`;

exports[`<RadioButton> Snapshots and structure should render defaults 1`] = `
<div
  className="lumx-radio-button lumx-radio-button--is-unchecked lumx-radio-button--theme-light"
>
  <div
    className="lumx-radio-button__input-wrapper"
  >
    <input
      className="lumx-radio-button__input-native"
      id="fixedId"
      onChange={[Function]}
      tabIndex={0}
      type="radio"
    />
    <div
      className="lumx-radio-button__input-placeholder"
    >
      <div
        className="lumx-radio-button__input-background"
      />
      <div
        className="lumx-radio-button__input-indicator"
      />
    </div>
  </div>
  <div
    className="lumx-radio-button__content"
  />
</div>
`;

exports[`<RadioButton> Snapshots and structure should render label & helper 1`] = `
<div
  className="lumx-radio-button lumx-radio-button--is-unchecked lumx-radio-button--theme-light"
>
  <div
    className="lumx-radio-button__input-wrapper"
  >
    <input
      className="lumx-radio-button__input-native"
      id="fixedId"
      onChange={[Function]}
      tabIndex={0}
      type="radio"
    />
    <div
      className="lumx-radio-button__input-placeholder"
    >
      <div
        className="lumx-radio-button__input-background"
      />
      <div
        className="lumx-radio-button__input-indicator"
      />
    </div>
  </div>
  <div
    className="lumx-radio-button__content"
  >
    <InputLabel
      className="lumx-radio-button__label"
      htmlFor="fixedId"
      theme="light"
    >
      Label
    </InputLabel>
    <InputHelper
      className="lumx-radio-button__helper"
      kind="info"
      theme="light"
    >
      Helper
    </InputHelper>
  </div>
</div>
`;
